include/algorithms/synthesis/syrec_synthesis.hpp

```diff
@@ -44,115 +35,149 @@ namespace syrec::internal {
         // shift operations
-        virtual void left_shift(const std::vector<unsigned>& dest, const std::vector<unsigned>& src1, unsigned src2);  // <<
-        virtual void right_shift(const std::vector<unsigned>& dest, const std::vector<unsigned>& src1, unsigned src2); // >>
+        void left_shift(const std::vector<unsigned>& dest, const std::vector<unsigned>& src1, unsigned src2);  // <<
+        void right_shift(const std::vector<unsigned>& dest, const std::vector<unsigned>& src1, unsigned src2); // >>
 
-        // efficient controls
-        virtual void add_active_control(unsigned);
-        virtual void remove_active_control(unsigned);
+        void add_active_control(unsigned);
+        void remove_active_control(unsigned);
 
-        virtual bool assemble_circuit(const cct_node&);
+        bool assemble_circuit(const cct_node&);
 
         cct_manager cct_man;
 
-        virtual void add_variable(circuit& circ, const std::vector<unsigned>& dimensions, const variable::ptr& var, constant _constant, bool _garbage, const std::string& arraystr);
-        virtual void get_variables(const variable_access::ptr& var, std::vector<unsigned>& lines);
+        void add_variable(circuit& circ, const std::vector<unsigned>& dimensions, const variable::ptr& var, constant _constant, bool _garbage, const std::string& arraystr);
+        void get_variables(const variable_access::ptr& var, std::vector<unsigned>& lines);
+
+        unsigned get_constant_line(bool value);
+        void     get_constant_lines(unsigned bitwidth, unsigned value, std::vector<unsigned>& lines);
 
-        virtual unsigned get_constant_line(bool value);
-        virtual void     get_constant_lines(unsigned bitwidth, unsigned value, std::vector<unsigned>& lines);
+        std::stack<statement::ptr>    _stmts;
+        circuit&                      _circ;
+        number::loop_variable_mapping loop_map;
+        std::stack<module::ptr>       modules;
 
     private:
-        circuit&                              _circ;
-        std::stack<statement::ptr>            _stmts;
         var_lines_map                         _var_lines;
         std::map<bool, std::vector<unsigned>> free_const_lines_map;
-        number::loop_variable_mapping         loop_map;
+    };
+
+    class SyrecSynthesisNoAdditionalLines: public SyrecSynthesis {
+    public:
+        using SyrecSynthesis::SyrecSynthesis;
+        ~SyrecSynthesisNoAdditionalLines() = default;
+
+        bool on_module(const module::ptr&);
+
+        bool full_statement(const statement::ptr& statement);
+        bool full_statement(const assign_statement& statement);
+
+        bool on_statement(const statement::ptr& statement);
+        bool on_statement(const assign_statement& statement);
+        bool on_statement(const if_statement& statement);
+        bool on_statement(const for_statement& statement);
+        bool on_statement(const call_statement& statement);
+        bool on_statement(const uncall_statement& statement);
```

Comment:
so for the functions to be name virtual in the base class, the declaration and the signature should be the same right?

Since they aren't, i decided to change the names